File: 백준/Silver/1149. RGB거리/RGB거리.js
const fs = require('fs');
const stdin = fs.readFileSync('/dev/stdin').toString().trim().split('\n');

const input = (() => {
  let line = 0;
  return () => stdin[line++];
})();

const n = parseInt(input());
const cost = Array.from({ length: n }, () => input().split(' ').map(Number));

const [R, G, B] = [0, 1, 2];

const dp = Array.from({ length: n }, () => new Array(3));
[dp[0][R], dp[0][G], dp[0][B]] = cost[0];

for (let i = 1; i < n; i++) {
  dp[i][R] = Math.min(dp[i - 1][G], dp[i - 1][B]) + cost[i][R];
  dp[i][G] = Math.min(dp[i - 1][R], dp[i - 1][B]) + cost[i][G];
  dp[i][B] = Math.min(dp[i - 1][G], dp[i - 1][R]) + cost[i][B];
}

const answer = Math.min(dp[n - 1][R], dp[n - 1][G], dp[n - 1][B]);

console.log(answer);

File: 백준/Silver/1149. RGB거리/RGB거리.py
import sys

sys.setrecursionlimit(10 ** 6)

n = int(input())
cost = [list(map(int, input().split())) for _ in range(n)]

available_color = [[1, 2], [0, 2], [0, 1]]
# memo[i][0] := i번쨰 집을 0번째 색으로 칠하는데 드는 비용
memo = [[-1] * 3 for _ in range(n)]


def dp(i, color):
    # base case
    if i == 0:
        return cost[i][color]

    # memoization
    if memo[i][color] != -1:
        return memo[i][color]

    memo[i][color] = cost[i][color]
    memo[i][color] += min(dp(i - 1, available_color[color][0]), dp(i - 1, available_color[color][1]))
    return memo[i][color]


print(min(dp(n - 1, 0), dp(n - 1, 1), dp(n - 1, 2)))
